storybookjs/storybook
# Core: StorybookConfig `stories`-field support type async function

Closes #21524 



## What I did

Modifies to the StorybookConfig['stories'] type to include a function that returns a Promise that resolves to `StoriesEntry[]`. This should cover the custom implementation case. Since this logic is already supported where presets are loaded/resolved and is already documented, I did not make changes to documentation or add tests. 

## How to test

1. modify `main.ts` in `ui/.storybook` so that the `stories` key in the config has an async function that resolves a promise like this: 

```
const asyncStories = Promise.resolve(allStories)
const config: StorybookConfig = {
     stories: async () => [... (await asyncStories)],
     // ...rest of the config
```

2. After this modification you should not see a typescript error on the stories key regardless of how you configure stories (async func or as a `StoriesEntry` array.
3. run yarn start to run the storybook dev server
4. all stories should load as expected.

## Checklist



- [x] Make sure to add/update documentation regarding your changes (updated docstring)
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

      make sure to add the `ci:merged` or `ci:daily` GH label to it.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`
